pilot/pkg/serviceregistry/serviceentry/controller.go

```diff
@@ -897,27 +891,66 @@ func autoAllocateIPs(services []*model.Service) []*model.Service {
 		// 3. the hostname is not a wildcard
 		if svc.DefaultAddress == constants.UnspecifiedIP && !svc.Hostname.IsWildCarded() &&
 			svc.Resolution != model.Passthrough {
-			n := svc.Hostname.String()
-			if v, ok := hnMap[n]; ok {
-				log.Debugf("Reuse IP for domain %s", n)
-
-				setAutoAllocatedIPs(svc, v)
+			hash.Write([]byte(svc.Attributes.Namespace + "/" + svc.Hostname.String()))
+			// First hash is calculated by
+			s := hash.Sum32()
+			fh := s % uint32(maxIPs)
+			// Check if there is no service with this hash first. If there is not service
+			// at this location - then we can safely assign this position for this service.
+			if hashedServices[fh] == nil {
+				hashedServices[fh] = svc
 			} else {
-				x++
-				if x%255 == 0 {
-					x++
+				// This means we have a collision. Resolve collision by "DoubleHashing".
+				i := uint32(1)
+				for {
+					sh := uint32(prime) - (s % uint32(prime))
+					nh := (s + i*sh) % uint32(maxIPs)
+					if hashedServices[nh] == nil {
+						hashedServices[nh] = svc
+						break
+					}
```

Comment:
Where is this rehash borrowed?

Reply:
Sorry. Did not understand. Are you asking - if this is borrowed from any other OSS? If not - it is not. This is a standard technique used for hash collisions and I implemented it.

Reply:
yes i meant that

Reply:
Consistent hash can reduce the influence of adding/deleting a service

Reply:
We will have to add all 65535 ips as nodes in the ring. Not sure how it will impact and also I am not 100% sure if it really helps here - it has to find the nearest based on "hash" function.